Stopped game only bw.setup join

Wenn ein BW Game gestoppt ist, (aus welchem grund auch immer) sollten nur Spieler mit der bw.setup permission joinen können.

```javascript
[21:39:36] [Server thread/ERROR]: Could not load 'plugins/BedwarsRel-1.2.8.jar' in folder 'plugins'
org.bukkit.plugin.InvalidPluginException: java.lang.UnsupportedClassVersionError: io/github/yannici/bedwars/Main : Unsupported major.minor version 52.0
	at org.bukkit.plugin.java.JavaPluginLoader.loadPlugin(JavaPluginLoader.java:135) ~[spigot.jar:git-Spigot-fdc1440-53fac9f]
	at org.bukkit.plugin.SimplePluginManager.loadPlugin(SimplePluginManager.java:329) ~[spigot.jar:git-Spigot-fdc1440-53fac9f]
	at org.bukkit.plugin.SimplePluginManager.loadPlugins(SimplePluginManager.java:251) [spigot.jar:git-Spigot-fdc1440-53fac9f]
	at org.bukkit.craftbukkit.v1_8_R3.CraftServer.loadPlugins(CraftServer.java:292) [spigot.jar:git-Spigot-fdc1440-53fac9f]
	at net.minecraft.server.v1_8_R3.DedicatedServer.init(DedicatedServer.java:198) [spigot.jar:git-Spigot-fdc1440-53fac9f]
	at net.minecraft.server.v1_8_R3.MinecraftServer.run(MinecraftServer.java:524) [spigot.jar:git-Spigot-fdc1440-53fac9f]
	at java.lang.Thread.run(Thread.java:745) [?:1.7.0_91]
Caused by: java.lang.UnsupportedClassVersionError: io/github/yannici/bedwars/Main : Unsupported major.minor version 52.0
	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.7.0_91]
	at java.lang.ClassLoader.defineClass(ClassLoader.java:800) ~[?:1.7.0_91]
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142) ~[?:1.7.0_91]
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:449) ~[?:1.7.0_91]
	at java.net.URLClassLoader.access$100(URLClassLoader.java:71) ~[?:1.7.0_91]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361) ~[?:1.7.0_91]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355) ~[?:1.7.0_91]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.7.0_91]
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354) ~[?:1.7.0_91]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:101) ~[spigot.jar:git-Spigot-fdc1440-53fac9f]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:86) ~[spigot.jar:git-Spigot-fdc1440-53fac9f]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425) ~[?:1.7.0_91]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358) ~[?:1.7.0_91]
	at java.lang.Class.forName0(Native Method) ~[?:1.7.0_91]
	at java.lang.Class.forName(Class.java:278) ~[?:1.7.0_91]
	at org.bukkit.plugin.java.PluginClassLoader.<init>(PluginClassLoader.java:64) ~[spigot.jar:git-Spigot-fdc1440-53fac9f]
	at org.bukkit.plugin.java.JavaPluginLoader.loadPlugin(JavaPluginLoader.java:131) ~[spigot.jar:git-Spigot-fdc1440-53fac9f]
	... 6 more
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,3 @@
 /bin/
 /dist
+/build

--- a/build.xml
+++ b/build.xml
@@ -3,6 +3,7 @@
         jenkins ant build file for Bedwars
     </description>
 	<!-- set global properties for this build -->
+	<property environment="env" />
 	<property name="src" location="./" />
 	<property name="build" location="build" />
 	<property name="dist" location="dist" />
@@ -41,7 +42,7 @@
 		<mkdir dir="${dist}" />
 
 		<!-- Put everything in ${build} into the BedwarsRel-${version}.jar file -->
-		<jar jarfile="${dist}/BedwarsRel-${version}-${GIT_COMMIT}.jar" basedir="${build}">
+		<jar jarfile="${dist}/BedwarsRel-${version}-${env.GIT_COMMIT}.jar" basedir="${build}">
 			<fileset dir="./">
 			    <include name="config.yml" />
 				<include name="plugin.yml" />

--- a/build.xml
+++ b/build.xml
@@ -41,7 +41,7 @@
 		<mkdir dir="${dist}" />
 
 		<!-- Put everything in ${build} into the BedwarsRel-${version}.jar file -->
-		<jar jarfile="${dist}/BedwarsRel-${version}.jar" basedir="${build}">
+		<jar jarfile="${dist}/BedwarsRel-${version}-${env.GIT_COMMIT}.jar" basedir="${build}">
 			<fileset dir="./">
 			    <include name="config.yml" />
 				<include name="plugin.yml" />

--- a/build.xml
+++ b/build.xml
@@ -41,7 +41,7 @@
 		<mkdir dir="${dist}" />
 
 		<!-- Put everything in ${build} into the BedwarsRel-${version}.jar file -->
-		<jar jarfile="${dist}/BedwarsRel-${version}-${env.GIT_COMMIT}.jar" basedir="${build}">
+		<jar jarfile="${dist}/BedwarsRel-${version}-${GIT_COMMIT}.jar" basedir="${build}">
 			<fileset dir="./">
 			    <include name="config.yml" />
 				<include name="plugin.yml" />

--- a/src/io/github/yannici/bedwars/Game/BungeeGameCycle.java
+++ b/src/io/github/yannici/bedwars/Game/BungeeGameCycle.java
@@ -164,7 +164,17 @@ public class BungeeGameCycle extends GameCycle {
 				}.runTaskLater(Main.getInstance(), 60L);
 			} else {
 				if (this.getGame().getState() == GameState.RUNNING && !Main.getInstance().spectationEnabled()) {
-					this.bungeeSendToServer(Main.getInstance().getBungeeHub(), p, false);
+					
+					new BukkitRunnable() {
+
+						@Override
+						public void run() {
+							BungeeGameCycle.this.bungeeSendToServer(Main.getInstance().getBungeeHub(), p, false);
+						}
+
+					}.runTaskLater(Main.getInstance(), 5L);
+					
+
 					new BukkitRunnable() {
 
 						@Override

--- a/src/io/github/yannici/bedwars/Game/BungeeGameCycle.java
+++ b/src/io/github/yannici/bedwars/Game/BungeeGameCycle.java
@@ -12,8 +12,10 @@ import io.github.yannici.bedwars.Events.BedwarsGameEndEvent;
 
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
-import org.bukkit.GameMode;
+import org.bukkit.Material;
 import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
 import org.bukkit.scheduler.BukkitRunnable;
 
 import com.google.common.collect.ImmutableMap;
@@ -217,39 +219,40 @@ public class BungeeGameCycle extends GameCycle {
 		}.runTaskLater(Main.getInstance(), (preventDelay) ? 0L : 20L);
 	}
 
-	@SuppressWarnings("deprecation")
 	@Override
 	public void onGameOver(GameOverTask task) {
 		if (Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true)) {
-			ArrayList<Player> players = new ArrayList<Player>();
+			final ArrayList<Player> players = new ArrayList<Player>();
+			final Game game = this.getGame();
 			players.addAll(this.getGame().getTeamPlayers());
 			players.addAll(this.getGame().getFreePlayers());
 			for (Player player : players) {
 
 				if (!player.getWorld().equals(this.getGame().getLobby().getWorld())) {
-					this.getGame().getPlayerSettings(player).setTeleporting(true);
+					game.getPlayerSettings(player).setTeleporting(true);
 					player.teleport(this.getGame().getLobby());
-
-					this.getGame().getPlayerStorage(player).clean();
-
-					// lobby gamemode
-					GameMode mode = GameMode.SURVIVAL;
-					try {
-						mode = GameMode.getByValue(Main.getInstance().getIntConfig("lobby-gamemode", 0));
-					} catch (Exception ex) {
-						// not valid gamemode
-					}
-
-					if (mode == null) {
-						mode = GameMode.SURVIVAL;
-					}
-					player.setGameMode(mode);
-
-					for (Player playerToShow : players) {
-						player.showPlayer(playerToShow);
-					}
+					game.getPlayerStorage(player).clean();					
 				}
 			}
+			
+			new BukkitRunnable() {
+				@Override
+				public void run() {				
+					for (Player player : players) {						
+						game.setPlayerGameMode(player);
+						game.setPlayerVisibility(player);
+						
+						// Leave Game (Slimeball)
+						ItemStack leaveGame = new ItemStack(Material.SLIME_BALL, 1);
+						ItemMeta im = leaveGame.getItemMeta();
+						im.setDisplayName(Main._l("lobby.leavegame"));
+						leaveGame.setItemMeta(im);
+						player.getInventory().setItem(8, leaveGame);
+						player.updateInventory();
+						
+					}	
+				}
+			}.runTaskLater(Main.getInstance(), 20L);
 		}
 		if (task.getCounter() == task.getStartCount() && task.getWinner() != null) {
 			this.getGame().broadcast(ChatColor.GOLD + Main._l("ingame.teamwon",

--- a/src/io/github/yannici/bedwars/Game/Game.java
+++ b/src/io/github/yannici/bedwars/Game/Game.java
@@ -1273,7 +1273,7 @@ public class Game {
 	}
 
 	public void removeTeam(Team team) {
-		this.teams.remove(team);
+		this.teams.remove(team.getName());
 		this.updateSigns();
 	}
 

--- a/src/io/github/yannici/bedwars/Game/Game.java
+++ b/src/io/github/yannici/bedwars/Game/Game.java
@@ -621,12 +621,30 @@ public class Game {
 			this.displayMapInfo(p);
 		} else {
 
-			Location location = this.getPlayerTeleportLocation(p);
+			final Location location = this.getPlayerTeleportLocation(p);
 
 			if (!p.getLocation().equals(location)) {
 				this.getPlayerSettings(p).setTeleporting(true);
-				p.teleport(location);
+				new BukkitRunnable() {
+
+					@Override
+					public void run() {
+						p.teleport(location);
+					}
+
+				}.runTaskLater(Main.getInstance(), 1L);
 			}
+			
+			
+			new BukkitRunnable() {
+
+				@Override
+				public void run() {
+					Game.this.setPlayerGameMode(p);
+					Game.this.setPlayerVisibility(p);
+				}
+
+			}.runTaskLater(Main.getInstance(), 2L);
 
 			this.broadcast(ChatColor.GREEN
 					+ Main._l("lobby.playerjoin", ImmutableMap.of("player", p.getDisplayName() + ChatColor.GREEN)));

--- a/src/io/github/yannici/bedwars/Game/Game.java
+++ b/src/io/github/yannici/bedwars/Game/Game.java
@@ -425,6 +425,32 @@ public class Game {
 			this.freePlayers.add(player);
 		}
 
+		final Location location = this.getPlayerTeleportLocation(p);
+
+		if (!p.getLocation().equals(location)) {
+			this.getPlayerSettings(p).setTeleporting(true);
+			new BukkitRunnable() {
+
+				@Override
+				public void run() {
+					p.teleport(location);
+				}
+
+			}.runTaskLater(Main.getInstance(), 1L);
+			
+		}
+
+		new BukkitRunnable() {
+
+			@Override
+			public void run() {
+				Game.this.setPlayerGameMode(p);
+				Game.this.setPlayerVisibility(p);
+			}
+
+		}.runTaskLater(Main.getInstance(), 2L);
+
+
 		PlayerStorage storage = this.getPlayerStorage(player);
 		if (storage != null) {
 			storage.clean();
@@ -433,31 +459,7 @@ public class Game {
 			storage.store();
 			storage.clean();
 		}
-
-		new BukkitRunnable() {
-
-			@Override
-			public void run() {
-				p.setAllowFlight(true);
-				p.setFlying(true);
-
-				// 1.7 compatible
-				try {
-					p.setGameMode(GameMode.valueOf("SPECTATOR"));
-				} catch (Exception ex) {
-					p.setGameMode(GameMode.SURVIVAL);
-				}
-
-				for (Player pl : Game.this.getPlayers()) {
-					if (pl.equals(p)) {
-						continue;
-					}
-
-					pl.hidePlayer(p);
-				}
-			}
-
-		}.runTaskLater(Main.getInstance(), 5L);
+		
 
 		// Leave Game (Slimeball)
 		ItemStack leaveGame = new ItemStack(Material.SLIME_BALL, 1);
@@ -465,6 +467,12 @@ public class Game {
 		im.setDisplayName(Main._l("lobby.leavegame"));
 		leaveGame.setItemMeta(im);
 		p.getInventory().setItem(8, leaveGame);
+		
+		if (this.getCycle() instanceof BungeeGameCycle && this.getCycle().isEndGameRunning()
+				&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true)) {
+			p.updateInventory();
+			return;
+		}
 
 		// Teleport to player (Compass)
 		ItemStack teleportPlayer = new ItemStack(Material.COMPASS, 1);
@@ -475,13 +483,100 @@ public class Game {
 
 		p.updateInventory();
 		this.updateScoreboard();
+
+	}
+
+	public Location getPlayerTeleportLocation(Player player) {
+		if (this.isSpectator(player)
+				&& !(this.getCycle() instanceof BungeeGameCycle && this.getCycle().isEndGameRunning()
+						&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true))) {
+			return ((Team) this.teams.values().toArray()[Utils.randInt(0, this.teams.size() - 1)]).getSpawnLocation();
+		}
+
+		if (this.getPlayerTeam(player) != null
+				&& !(this.getCycle() instanceof BungeeGameCycle && this.getCycle().isEndGameRunning()
+						&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true))) {
+			return this.getPlayerTeam(player).getSpawnLocation();
+		}
+
+		return this.getLobby();
+	}
+
+	@SuppressWarnings("deprecation")
+	public void setPlayerGameMode(Player player) {
+		if (this.isSpectator(player)
+				&& !(this.getCycle() instanceof BungeeGameCycle && this.getCycle().isEndGameRunning()
+						&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true))) {
+
+			player.setAllowFlight(true);
+			player.setFlying(true);
+
+			// 1.7 compatible
+			try {
+				player.setGameMode(GameMode.valueOf("SPECTATOR"));
+			} catch (Exception ex) {
+				player.setGameMode(GameMode.SURVIVAL);
+			}
+		} else {
+			GameMode gameMode = null;
+			try {
+				gameMode = GameMode.getByValue(Main.getInstance().getIntConfig("lobby-gamemode", 0));
+			} catch (Exception ex) {
+				// not valid gamemode
+			}
+
+			if (gameMode == null) {
+				gameMode = GameMode.SURVIVAL;
+			}
+			player.setGameMode(gameMode);
+		}
+	}
+
+	public void setPlayerVisibility(Player player) {
+		ArrayList<Player> players = new ArrayList<Player>();
+		players.addAll(this.getPlayers());
+
+		if (this.state == GameState.RUNNING
+				&& !(this.getCycle() instanceof BungeeGameCycle && this.getCycle().isEndGameRunning()
+						&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true))) {
+			if (this.isSpectator(player)) {
+				if (player.getGameMode().equals(GameMode.SURVIVAL)) {
+					for (Player playerInGame : players) {
+						playerInGame.hidePlayer(player);
+						player.showPlayer(playerInGame);
+					}
+				} else {
+					for (Player teamPlayer : this.getTeamPlayers()) {
+						teamPlayer.hidePlayer(player);
+						player.showPlayer(teamPlayer);
+					}
+					for (Player freePlayer : this.getFreePlayers()) {
+						freePlayer.showPlayer(player);
+						player.showPlayer(freePlayer);
+					}
+				}
+			} else {
+				for (Player playerInGame : players) {
+					playerInGame.showPlayer(player);
+					player.showPlayer(playerInGame);
+				}
+			}
+		} else {
+			for (Player playerInGame : players) {
+				if (!playerInGame.equals(player)) {
+					playerInGame.showPlayer(player);
+					player.showPlayer(playerInGame);
+				}
+			}
+		}
+
 	}
 
 	public boolean isSpectator(Player player) {
 		return (this.getState() == GameState.RUNNING && this.freePlayers.contains(player));
 	}
 
-	public boolean playerJoins(Player p) {
+	public boolean playerJoins(final Player p) {
 		if (this.state == GameState.STOPPED
 				|| (this.state == GameState.RUNNING && !Main.getInstance().spectationEnabled())) {
 			if (this.cycle instanceof BungeeGameCycle) {
@@ -516,14 +611,23 @@ public class Game {
 		// add player settings
 		this.addPlayerSettings(p);
 
+		for (Player playerInGame : this.getPlayers()) {
+			playerInGame.hidePlayer(p);
+			p.hidePlayer(playerInGame);
+		}
+
 		if (this.state == GameState.RUNNING) {
 			this.toSpectator(p);
-
-			this.getPlayerSettings(p).setTeleporting(true);
-			p.teleport(
-					((Team) this.teams.values().toArray()[Utils.randInt(0, this.teams.size() - 1)]).getSpawnLocation());
 			this.displayMapInfo(p);
 		} else {
+
+			Location location = this.getPlayerTeleportLocation(p);
+
+			if (!p.getLocation().equals(location)) {
+				this.getPlayerSettings(p).setTeleporting(true);
+				p.teleport(location);
+			}
+
 			this.broadcast(ChatColor.GREEN
 					+ Main._l("lobby.playerjoin", ImmutableMap.of("player", p.getDisplayName() + ChatColor.GREEN)));
 
@@ -538,10 +642,6 @@ public class Game {
 			storage.store();
 			storage.clean();
 
-			if (!this.lobby.getWorld().equals(p.getWorld())) {
-				this.getPlayerSettings(p).setTeleporting(true);
-			}
-			p.teleport(this.lobby);
 			storage.loadLobbyInventory(this);
 
 			if (Main.getInstance().getBooleanConfig("store-game-records", true)) {
@@ -559,8 +659,8 @@ public class Game {
 					}
 				} else {
 					int playersNeeded = this.getMinPlayers() - this.getPlayerAmount();
-					this.broadcast(ChatColor.GREEN
-							+ Main._l("lobby.moreplayersneeded", "count", ImmutableMap.of("count", String.valueOf(playersNeeded))));
+					this.broadcast(ChatColor.GREEN + Main._l("lobby.moreplayersneeded", "count",
+							ImmutableMap.of("count", String.valueOf(playersNeeded))));
 
 				}
 			}
@@ -572,6 +672,7 @@ public class Game {
 		this.updateScoreboard();
 		this.updateSigns();
 		return true;
+
 	}
 
 	public boolean playerLeave(Player p, boolean kicked) {

--- a/src/io/github/yannici/bedwars/Game/Game.java
+++ b/src/io/github/yannici/bedwars/Game/Game.java
@@ -427,7 +427,7 @@ public class Game {
 
 		final Location location = this.getPlayerTeleportLocation(p);
 
-		if (!p.getLocation().equals(location)) {
+		if (!p.getLocation().getWorld().equals(location.getWorld())) {
 			this.getPlayerSettings(p).setTeleporting(true);
 			new BukkitRunnable() {
 
@@ -436,8 +436,8 @@ public class Game {
 					p.teleport(location);
 				}
 
-			}.runTaskLater(Main.getInstance(), 1L);
-			
+			}.runTaskLater(Main.getInstance(), 10L);
+
 		}
 
 		new BukkitRunnable() {
@@ -448,8 +448,7 @@ public class Game {
 				Game.this.setPlayerVisibility(p);
 			}
 
-		}.runTaskLater(Main.getInstance(), 2L);
-
+		}.runTaskLater(Main.getInstance(), 15L);
 
 		PlayerStorage storage = this.getPlayerStorage(player);
 		if (storage != null) {
@@ -459,7 +458,6 @@ public class Game {
 			storage.store();
 			storage.clean();
 		}
-		
 
 		// Leave Game (Slimeball)
 		ItemStack leaveGame = new ItemStack(Material.SLIME_BALL, 1);
@@ -467,7 +465,7 @@ public class Game {
 		im.setDisplayName(Main._l("lobby.leavegame"));
 		leaveGame.setItemMeta(im);
 		p.getInventory().setItem(8, leaveGame);
-		
+
 		if (this.getCycle() instanceof BungeeGameCycle && this.getCycle().isEndGameRunning()
 				&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true)) {
 			p.updateInventory();
@@ -611,31 +609,39 @@ public class Game {
 		// add player settings
 		this.addPlayerSettings(p);
 
-		for (Player playerInGame : this.getPlayers()) {
-			playerInGame.hidePlayer(p);
-			p.hidePlayer(playerInGame);
-		}
+		new BukkitRunnable() {
+
+			@Override
+			public void run() {
+				for (Player playerInGame : Game.this.getPlayers()) {
+					playerInGame.hidePlayer(p);
+					p.hidePlayer(playerInGame);
+				}
+			}
+
+		}.runTaskLater(Main.getInstance(), 5L);
 
 		if (this.state == GameState.RUNNING) {
 			this.toSpectator(p);
 			this.displayMapInfo(p);
 		} else {
 
-			final Location location = this.getPlayerTeleportLocation(p);
+			if (!Utils.isSupportingTitles()) {
+				final Location location = this.getPlayerTeleportLocation(p);
 
-			if (!p.getLocation().equals(location)) {
-				this.getPlayerSettings(p).setTeleporting(true);
-				new BukkitRunnable() {
+				if (!p.getLocation().equals(location)) {
+					this.getPlayerSettings(p).setTeleporting(true);
+					new BukkitRunnable() {
 
-					@Override
-					public void run() {
-						p.teleport(location);
-					}
+						@Override
+						public void run() {
+							p.teleport(location);
+						}
 
-				}.runTaskLater(Main.getInstance(), 1L);
+					}.runTaskLater(Main.getInstance(), 10L);
+				}
 			}
-			
-			
+
 			new BukkitRunnable() {
 
 				@Override
@@ -644,7 +650,7 @@ public class Game {
 					Game.this.setPlayerVisibility(p);
 				}
 
-			}.runTaskLater(Main.getInstance(), 2L);
+			}.runTaskLater(Main.getInstance(), 15L);
 
 			this.broadcast(ChatColor.GREEN
 					+ Main._l("lobby.playerjoin", ImmutableMap.of("player", p.getDisplayName() + ChatColor.GREEN)));

--- a/src/io/github/yannici/bedwars/Game/PlayerStorage.java
+++ b/src/io/github/yannici/bedwars/Game/PlayerStorage.java
@@ -140,7 +140,6 @@ public class PlayerStorage {
         return this.left;
     }
 
-    @SuppressWarnings("deprecation")
     public void loadLobbyInventory(Game game) {
     	ItemMeta im = null;
     	
@@ -182,20 +181,6 @@ public class PlayerStorage {
             this.player.getInventory().addItem(startGame);
         }
 
-        
-        // lobby gamemode
-        GameMode mode = GameMode.SURVIVAL;
-        try {
-            mode = GameMode.getByValue(Main.getInstance().getIntConfig("lobby-gamemode", 0));
-        } catch(Exception ex) {
-            // not valid gamemode
-        }
-        
-        if(mode == null) {
-            mode = GameMode.SURVIVAL;
-        }
-        
-        this.player.setGameMode(mode);
         this.player.updateInventory();
     }
 

--- a/src/io/github/yannici/bedwars/Listener/HangingListener.java
+++ b/src/io/github/yannici/bedwars/Listener/HangingListener.java
@@ -1,0 +1,26 @@
+package io.github.yannici.bedwars.Listener;
+
+import org.bukkit.entity.Hanging;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.hanging.HangingBreakEvent;
+import org.bukkit.event.hanging.HangingBreakEvent.RemoveCause;
+
+public class HangingListener extends BaseListener {
+	
+	public HangingListener() {
+		super();
+	}
+
+	@EventHandler(priority = EventPriority.HIGHEST)
+	public void onHangingBreak(HangingBreakEvent event) {
+		Hanging hanging = event.getEntity();
+		if (event.getCause().equals(RemoveCause.OBSTRUCTION)) {
+			hanging.getLocation().getBlock().breakNaturally();
+			event.setCancelled(true);
+		} else if (event.getCause().equals(RemoveCause.EXPLOSION)) {
+			event.setCancelled(true);
+		}
+
+	}
+}

--- a/src/io/github/yannici/bedwars/Listener/Player18Listener.java
+++ b/src/io/github/yannici/bedwars/Listener/Player18Listener.java
@@ -1,0 +1,29 @@
+package io.github.yannici.bedwars.Listener;
+
+import java.util.ArrayList;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.spigotmc.event.player.PlayerSpawnLocationEvent;
+
+import io.github.yannici.bedwars.Main;
+import io.github.yannici.bedwars.Game.Game;
+
+public class Player18Listener extends BaseListener {
+    
+    @EventHandler(priority = EventPriority.HIGH)
+    public void onPlayerSpawnLocation(PlayerSpawnLocationEvent event) {
+    	Player player = event.getPlayer();
+    	
+    	ArrayList<Game> games = Main.getInstance().getGameManager().getGames();
+		if (games.size() == 0) {
+			return;
+		}
+
+		Game firstGame = games.get(0);
+		
+        event.setSpawnLocation(firstGame.getPlayerTeleportLocation(player));
+    }
+
+}

--- a/src/io/github/yannici/bedwars/Listener/PlayerListener.java
+++ b/src/io/github/yannici/bedwars/Listener/PlayerListener.java
@@ -86,10 +86,17 @@ public class PlayerListener extends BaseListener {
 			}
 
 			if (!firstGame.playerJoins(player)) {
-				if (firstGame.getCycle() instanceof BungeeGameCycle) {
-					((BungeeGameCycle) firstGame.getCycle()).bungeeSendToServer(Main.getInstance().getBungeeHub(),
-							player, true);
-				}
+				new BukkitRunnable() {
+
+					@Override
+					public void run() {
+						if (firstGame.getCycle() instanceof BungeeGameCycle) {
+							((BungeeGameCycle) firstGame.getCycle())
+									.bungeeSendToServer(Main.getInstance().getBungeeHub(), player, true);
+						}
+					}
+
+				}.runTaskLater(Main.getInstance(), 5L);
 			}
 
 		}
@@ -535,11 +542,11 @@ public class PlayerListener extends BaseListener {
 			}
 
 			ce.setFormat(format);
-			
+
 			if (!Main.getInstance().getBooleanConfig("seperate-game-chat", true)) {
 				return;
 			}
-			
+
 			Iterator<Player> recipiens = ce.getRecipients().iterator();
 			while (recipiens.hasNext()) {
 				Player recipient = recipiens.next();
@@ -547,7 +554,7 @@ public class PlayerListener extends BaseListener {
 					recipiens.remove();
 				}
 			}
-			
+
 			return;
 		}
 

--- a/src/io/github/yannici/bedwars/Listener/PlayerListener.java
+++ b/src/io/github/yannici/bedwars/Listener/PlayerListener.java
@@ -100,7 +100,7 @@ public class PlayerListener extends BaseListener {
 					firstGame.setPlayerVisibility(player);
 				}
 
-			}.runTaskLater(Main.getInstance(), 1L);
+			}.runTaskLater(Main.getInstance(), 5L);
 		}
 
 		if (Main.getInstance().isHologramsEnabled() && Main.getInstance().getHolographicInteractor() != null) {

--- a/src/io/github/yannici/bedwars/Listener/PlayerListener.java
+++ b/src/io/github/yannici/bedwars/Listener/PlayerListener.java
@@ -92,15 +92,6 @@ public class PlayerListener extends BaseListener {
 				}
 			}
 
-			new BukkitRunnable() {
-
-				@Override
-				public void run() {
-					firstGame.setPlayerGameMode(player);
-					firstGame.setPlayerVisibility(player);
-				}
-
-			}.runTaskLater(Main.getInstance(), 5L);
 		}
 
 		if (Main.getInstance().isHologramsEnabled() && Main.getInstance().getHolographicInteractor() != null) {

--- a/src/io/github/yannici/bedwars/Listener/PlayerListener.java
+++ b/src/io/github/yannici/bedwars/Listener/PlayerListener.java
@@ -78,8 +78,8 @@ public class PlayerListener extends BaseListener {
 				return;
 			}
 
-			Player player = je.getPlayer();
-			Game firstGame = games.get(0);
+			final Player player = je.getPlayer();
+			final Game firstGame = games.get(0);
 
 			if (firstGame.getState() == GameState.STOPPED) {
 				return;
@@ -91,6 +91,16 @@ public class PlayerListener extends BaseListener {
 							player, true);
 				}
 			}
+
+			new BukkitRunnable() {
+
+				@Override
+				public void run() {
+					firstGame.setPlayerGameMode(player);
+					firstGame.setPlayerVisibility(player);
+				}
+
+			}.runTaskLater(Main.getInstance(), 1L);
 		}
 
 		if (Main.getInstance().isHologramsEnabled() && Main.getInstance().getHolographicInteractor() != null) {
@@ -366,7 +376,10 @@ public class PlayerListener extends BaseListener {
 	@SuppressWarnings({ "rawtypes", "unchecked" })
 	private void onIngameInventoryClick(InventoryClickEvent ice, Player player, Game game) {
 		if (!ice.getInventory().getName().equals(Main._l("ingame.shop.name"))) {
-			if (game.isSpectator(player)) {
+			if (game.isSpectator(player)
+					|| (game.getCycle() instanceof BungeeGameCycle && game.getCycle().isEndGameRunning()
+							&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true))) {
+
 				ItemStack clickedStack = ice.getCurrentItem();
 				if (clickedStack == null) {
 					return;
@@ -569,7 +582,7 @@ public class PlayerListener extends BaseListener {
 			}
 
 			ce.setFormat(format);
-			
+
 			if (!Main.getInstance().isBungee() || seperateSpectatorChat) {
 				Iterator<Player> recipiens = ce.getRecipients().iterator();
 				while (recipiens.hasNext()) {
@@ -814,7 +827,8 @@ public class PlayerListener extends BaseListener {
 				}
 			}
 
-			if (g.isSpectator(player)) {
+			if (g.isSpectator(player) || (g.getCycle() instanceof BungeeGameCycle && g.getCycle().isEndGameRunning()
+					&& Main.getInstance().getBooleanConfig("bungeecord.endgame-in-lobby", true))) {
 				if (interactingMaterial == Material.SLIME_BALL) {
 					g.playerLeave(player, false);
 					return;

--- a/src/io/github/yannici/bedwars/Listener/PlayerListener.java
+++ b/src/io/github/yannici/bedwars/Listener/PlayerListener.java
@@ -81,7 +81,7 @@ public class PlayerListener extends BaseListener {
 			final Player player = je.getPlayer();
 			final Game firstGame = games.get(0);
 
-			if (firstGame.getState() == GameState.STOPPED) {
+			if (firstGame.getState() == GameState.STOPPED && player.hasPermission("bw.setup")) {
 				return;
 			}
 

--- a/src/io/github/yannici/bedwars/Main.java
+++ b/src/io/github/yannici/bedwars/Main.java
@@ -13,6 +13,7 @@ import io.github.yannici.bedwars.Listener.ChunkListener;
 import io.github.yannici.bedwars.Listener.Entity18Listener;
 import io.github.yannici.bedwars.Listener.EntityListener;
 import io.github.yannici.bedwars.Listener.HangingListener;
+import io.github.yannici.bedwars.Listener.Player18Listener;
 import io.github.yannici.bedwars.Listener.PlayerListener;
 import io.github.yannici.bedwars.Listener.ServerListener;
 import io.github.yannici.bedwars.Listener.SignListener;
@@ -641,10 +642,16 @@ public class Main extends JavaPlugin {
 		new WeatherListener();
 		new BlockListener();
 		new PlayerListener();
+		
+		if (Utils.isSupportingTitles()) {
+			new Player18Listener();
+		}
+		
 		EntityListener el = new EntityListener();
 		if (Utils.isSupportingTitles()) {
 			new Entity18Listener(el);
 		}
+		
 		new HangingListener();
 		new ServerListener();
 		new SignListener();

--- a/src/io/github/yannici/bedwars/Main.java
+++ b/src/io/github/yannici/bedwars/Main.java
@@ -12,6 +12,7 @@ import io.github.yannici.bedwars.Listener.BlockListener;
 import io.github.yannici.bedwars.Listener.ChunkListener;
 import io.github.yannici.bedwars.Listener.Entity18Listener;
 import io.github.yannici.bedwars.Listener.EntityListener;
+import io.github.yannici.bedwars.Listener.HangingListener;
 import io.github.yannici.bedwars.Listener.PlayerListener;
 import io.github.yannici.bedwars.Listener.ServerListener;
 import io.github.yannici.bedwars.Listener.SignListener;
@@ -59,7 +60,7 @@ import com.google.common.collect.ImmutableMap;
 public class Main extends JavaPlugin {
 
 	private static Main instance = null;
-	
+
 	public static int PROJECT_ID = 91743;
 
 	private ArrayList<BaseCommand> commands = new ArrayList<BaseCommand>();
@@ -70,61 +71,61 @@ public class Main extends JavaPlugin {
 	private LocalizationConfig localization = null;
 	private DatabaseManager dbManager = null;
 	private BukkitTask updateChecker = null;
-	
+
 	private List<Material> breakableTypes = null;
 	private YamlConfiguration shopConfig = null;
-	
+
 	private HolographicDisplaysInteraction holographicInteraction = null;
-	
+
 	private boolean isSpigot = false;
-	
+
 	private static Boolean locationSerializable = null;
-	
+
 	private PlayerStatisticManager playerStatisticManager = null;
-	
+
 	private ScoreboardManager scoreboardManager = null;
 	private GameManager gameManager = null;
 
-    @Override
+	@Override
 	public void onEnable() {
 		Main.instance = this;
-		
+
 		// register classes
 		this.registerConfigurationClasses();
-		
+
 		// save default config
 		this.saveDefaultConfig();
 		this.loadConfigInUTF();
 
 		this.getConfig().options().copyDefaults(true);
 		this.getConfig().options().copyHeader(true);
-		
+
 		ConfigUpdater configUpdater = new ConfigUpdater();
 		configUpdater.addConfigs();
 		this.saveConfiguration();
-	    this.loadConfigInUTF();
-	    this.loadShop();
-		
-	    this.isSpigot = this.getIsSpigot();
+		this.loadConfigInUTF();
+		this.loadShop();
+
+		this.isSpigot = this.getIsSpigot();
 		this.loadDatabase();
-		
+
 		this.craftbukkit = this.getCraftBukkit();
 		this.minecraft = this.getMinecraftPackage();
 		this.version = this.loadVersion();
 
 		this.registerCommands();
 		this.registerListener();
-		
+
 		this.gameManager = new GameManager();
-		
+
 		// bungeecord
-        if(Main.getInstance().isBungee()) {
-            this.getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");
-        }
-		
+		if (Main.getInstance().isBungee()) {
+			this.getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");
+		}
+
 		this.loadStatistics();
 		this.localization = this.loadLocalization();
-		
+
 		this.checkUpdates();
 
 		// Loading
@@ -132,68 +133,68 @@ public class Main extends JavaPlugin {
 		this.gameManager.loadGames();
 		this.startTimeListener();
 		this.startMetricsIfEnabled();
-		
+
 		// holograms
-		if(this.isHologramsEnabled()) {
-		    this.holographicInteraction = new HolographicDisplaysInteraction();
-	        this.holographicInteraction.loadHolograms();
+		if (this.isHologramsEnabled()) {
+			this.holographicInteraction = new HolographicDisplaysInteraction();
+			this.holographicInteraction.loadHolograms();
 		}
 	}
-	
+
 	@Override
 	public void onDisable() {
 		this.stopTimeListener();
 		this.gameManager.unloadGames();
 		this.cleanDatabase();
-		
-		if(this.isHologramsEnabled() && this.holographicInteraction != null) {
-		    this.holographicInteraction.unloadHolograms();
+
+		if (this.isHologramsEnabled() && this.holographicInteraction != null) {
+			this.holographicInteraction.unloadHolograms();
 		}
 	}
 
 	public void loadConfigInUTF() {
-        File configFile = new File(this.getDataFolder(), "config.yml");
-        if(!configFile.exists()) {
-            return;
-        }
-        
-        try {
-            BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(configFile), "UTF-8"));
-            this.getConfig().load(reader);
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
-        
-        if(this.getConfig() == null) {
-        	return;
-        }
-        
-        // load breakable materials
-        this.breakableTypes = new ArrayList<Material>();
-        for(String material : this.getConfig().getStringList("breakable-blocks")) {
-        	if(material.equalsIgnoreCase("none")) {
-        		continue;
-        	}
+		File configFile = new File(this.getDataFolder(), "config.yml");
+		if (!configFile.exists()) {
+			return;
+		}
 
-        	Material mat = Utils.parseMaterial(material);
-        	if(mat == null) {
-        		continue;
-        	}
-        	
-        	if(this.breakableTypes.contains(mat)) {
-        		continue;
-        	}
-        	
-        	this.breakableTypes.add(mat);
-        }
-    }
-	
+		try {
+			BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(configFile), "UTF-8"));
+			this.getConfig().load(reader);
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+
+		if (this.getConfig() == null) {
+			return;
+		}
+
+		// load breakable materials
+		this.breakableTypes = new ArrayList<Material>();
+		for (String material : this.getConfig().getStringList("breakable-blocks")) {
+			if (material.equalsIgnoreCase("none")) {
+				continue;
+			}
+
+			Material mat = Utils.parseMaterial(material);
+			if (mat == null) {
+				continue;
+			}
+
+			if (this.breakableTypes.contains(mat)) {
+				continue;
+			}
+
+			this.breakableTypes.add(mat);
+		}
+	}
+
 	public void loadShop() {
 		File file = new File(Main.getInstance().getDataFolder(), "shop.yml");
-		if(!file.exists()) {
+		if (!file.exists()) {
 			// create default file
 			this.saveResource("shop.yml", false);
-			
+
 			// wait until it's really saved
 			try {
 				Thread.sleep(100);
@@ -201,133 +202,137 @@ public class Main extends JavaPlugin {
 				e.printStackTrace();
 			}
 		}
-		
+
 		this.shopConfig = new YamlConfiguration();
-		
+
 		try {
-            BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file), "UTF-8"));
-            this.shopConfig.load(reader);
-        } catch (Exception e) {
-        	this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.RED + "Couldn't load shop! Error in parsing shop!"));
-            e.printStackTrace();
-        }
+			BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file), "UTF-8"));
+			this.shopConfig.load(reader);
+		} catch (Exception e) {
+			this.getServer().getConsoleSender().sendMessage(
+					ChatWriter.pluginMessage(ChatColor.RED + "Couldn't load shop! Error in parsing shop!"));
+			e.printStackTrace();
+		}
 	}
-	
+
 	public void dispatchRewardCommands(List<String> commands, Map<String, String> replacements) {
-		for(String command : commands) {
+		for (String command : commands) {
 			command = command.trim();
-			if(command.equals("")) {
+			if (command.equals("")) {
 				continue;
 			}
-			
-			if(command.equalsIgnoreCase("none")) {
+
+			if (command.equalsIgnoreCase("none")) {
 				break;
 			}
-			
-			if(command.startsWith("/")) {
+
+			if (command.startsWith("/")) {
 				command = command.substring(1);
 			}
-			
-			for(Entry<String, String> entry : replacements.entrySet()) {
+
+			for (Entry<String, String> entry : replacements.entrySet()) {
 				command = command.replace(entry.getKey(), entry.getValue());
 			}
-			
+
 			Main.getInstance().getServer().dispatchCommand(Main.getInstance().getServer().getConsoleSender(), command);
 		}
 	}
-	
+
 	public void saveConfiguration() {
 		File file = new File(Main.getInstance().getDataFolder(), "config.yml");
 		try {
 			file.mkdirs();
-			
-			String data = this.getYamlDump((YamlConfiguration)this.getConfig());
-			
+
+			String data = this.getYamlDump((YamlConfiguration) this.getConfig());
+
 			FileOutputStream stream = new FileOutputStream(file);
 			OutputStreamWriter writer = new OutputStreamWriter(stream, "UTF-8");
-			
+
 			try {
 				writer.write(data);
 			} finally {
 				writer.close();
 				stream.close();
 			}
-		} catch(Exception ex) {
+		} catch (Exception ex) {
 			ex.printStackTrace();
 		}
 	}
-	
+
 	public Class<?> getVersionRelatedClass(String className) {
 		try {
-			Class<?> clazz = Class.forName("io.github.yannici.bedwars.Com." + this.getCurrentVersion() + "." + className);
+			Class<?> clazz = Class
+					.forName("io.github.yannici.bedwars.Com." + this.getCurrentVersion() + "." + className);
 			return clazz;
-		} catch(Exception ex) {
-			this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.RED + "Couldn't find version related class io.github.yannici.bedwars.Com." + this.getCurrentVersion() + "." + className));
+		} catch (Exception ex) {
+			this.getServer().getConsoleSender()
+					.sendMessage(ChatWriter.pluginMessage(
+							ChatColor.RED + "Couldn't find version related class io.github.yannici.bedwars.Com."
+									+ this.getCurrentVersion() + "." + className));
 		}
-		
+
 		return null;
 	}
-	
+
 	public String getYamlDump(YamlConfiguration config) {
 		try {
 			String fullstring = config.saveToString();
 			String endstring = fullstring;
 			endstring = Utils.unescape_perl_string(fullstring);
-			
+
 			return endstring;
-		} catch(Exception ex) {
+		} catch (Exception ex) {
 			ex.printStackTrace();
 		}
-		
+
 		return null;
 	}
-	
+
 	public boolean isBreakableType(Material type) {
 		return (this.breakableTypes.contains(type));
 	}
-	
+
 	public boolean isMineshafterPresent() {
-        try {
-            Class.forName("mineshafter.MineServer");
-            return true;
-        } catch (Exception e) {
-            return false;
-        }
-    }
-	
-	public PlayerStatisticManager getPlayerStatisticManager() {
-	    return this.playerStatisticManager;
+		try {
+			Class.forName("mineshafter.MineServer");
+			return true;
+		} catch (Exception e) {
+			return false;
+		}
 	}
-	
+
+	public PlayerStatisticManager getPlayerStatisticManager() {
+		return this.playerStatisticManager;
+	}
+
 	private void checkUpdates() {
 		try {
-			if(this.getBooleanConfig("check-updates", true)) {
+			if (this.getBooleanConfig("check-updates", true)) {
 				this.updateChecker = new BukkitRunnable() {
-                    
-                    @Override
-                    public void run() {
-                    	final BukkitRunnable task = this;
-                    	UpdateCallback callback = new UpdateCallback() {
-							
+
+					@Override
+					public void run() {
+						final BukkitRunnable task = this;
+						UpdateCallback callback = new UpdateCallback() {
+
 							@Override
 							public void onFinish(PluginUpdater updater) {
-								if(updater.getResult() == UpdateResult.SUCCESS) {
+								if (updater.getResult() == UpdateResult.SUCCESS) {
 									task.cancel();
 								}
 							}
 						};
-						
-                        new PluginUpdater(Main.getInstance(), Main.PROJECT_ID, 
-                        		Main.getInstance().getFile(), 
-                        		PluginUpdater.UpdateType.DEFAULT,
-                        		callback,
-                        		Main.getInstance().getBooleanConfig("update-infos", true));
-                    }
-                    
-                }.runTaskTimerAsynchronously(Main.getInstance(), 40L, 36000L);
+
+						new PluginUpdater(Main.getInstance(), Main.PROJECT_ID, Main.getInstance().getFile(),
+								PluginUpdater.UpdateType.DEFAULT, callback,
+								Main.getInstance().getBooleanConfig("update-infos", true));
+					}
+
+				}.runTaskTimerAsynchronously(Main.getInstance(), 40L, 36000L);
 			}
-		} catch(Exception ex) {
-			this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.RED + "Check for updates not successful: Error!"));
+		} catch (Exception ex) {
+			this.getServer().getConsoleSender()
+					.sendMessage(ChatWriter.pluginMessage(ChatColor.RED + "Check for updates not successful: Error!"));
 		}
 	}
 
@@ -338,97 +343,93 @@ public class Main extends JavaPlugin {
 		config.loadLocale(this.getConfig().getString("locale"), false);
 		return config;
 	}
-	
+
 	private void loadStatistics() {
 		this.playerStatisticManager = new PlayerStatisticManager();
 		this.playerStatisticManager.initialize();
 	}
-	
+
 	private void loadDatabase() {
-		if(!this.getBooleanConfig("statistics.enabled", false)
-				|| !this.getStringConfig("statistics.storage","yaml").equals("database")) {
+		if (!this.getBooleanConfig("statistics.enabled", false)
+				|| !this.getStringConfig("statistics.storage", "yaml").equals("database")) {
 			return;
 		}
-		
-		this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.GREEN + "Initialize database ..."));
+
+		this.getServer().getConsoleSender()
+				.sendMessage(ChatWriter.pluginMessage(ChatColor.GREEN + "Initialize database ..."));
 		this.loadingRequiredLibs();
-		
+
 		String host = this.getStringConfig("database.host", null);
 		int port = this.getIntConfig("database.port", 3306);
 		String user = this.getStringConfig("database.user", null);
 		String password = this.getStringConfig("database.password", null);
 		String db = this.getStringConfig("database.db", null);
-		
-		if(host == null || user == null || password == null || db == null) {
+
+		if (host == null || user == null || password == null || db == null) {
 			return;
 		}
-		
+
 		this.dbManager = new DatabaseManager(host, port, user, password, db);
 		this.dbManager.initialize();
-		
-		this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.GREEN + "Update database ..."));
+
+		this.getServer().getConsoleSender()
+				.sendMessage(ChatWriter.pluginMessage(ChatColor.GREEN + "Update database ..."));
 		(new DatabaseUpdater()).execute();
-		
+
 		this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.GREEN + "Done."));
 	}
-	
+
 	public StorageType getStatisticStorageType() {
 		String storage = this.getStringConfig("statistics.storage", "yaml");
 		return StorageType.getByName(storage);
 	}
-	
+
 	public boolean statisticsEnabled() {
 		return this.getBooleanConfig("statistics.enabled", false);
 	}
-	
+
 	private void cleanDatabase() {
-		if(this.dbManager != null) {
+		if (this.dbManager != null) {
 			this.dbManager.cleanUp();
 		}
 	}
-	
+
 	private void loadingRequiredLibs() {
 		try {
-            final File[] libs = new File[] {
-                    new File(this.getDataFolder() + "/lib/", "c3p0-0.9.5.jar"),
-                    new File(this.getDataFolder() + "/lib/", "mchange-commons-java-0.2.9.jar")};
-            for (final File lib : libs) {
-                if (!lib.exists()) {
-                    JarUtils.extractFromJar(lib.getName(),
-                            lib.getAbsolutePath());
-                }
-            }
-            for (final File lib : libs) {
-                if (!lib.exists()) {
-                    this.getLogger().warning(
-                            "There was a critical error loading bedwars plugin! Could not find lib: "
-                                    + lib.getName());
-                    Bukkit.getServer().getPluginManager().disablePlugin(this);
-                    return;
-                }
-                this.addClassPath(JarUtils.getJarUrl(lib));
-            }
-        } catch (final Exception e) {
-            e.printStackTrace();
-        }
+			final File[] libs = new File[] { new File(this.getDataFolder() + "/lib/", "c3p0-0.9.5.jar"),
+					new File(this.getDataFolder() + "/lib/", "mchange-commons-java-0.2.9.jar") };
+			for (final File lib : libs) {
+				if (!lib.exists()) {
+					JarUtils.extractFromJar(lib.getName(), lib.getAbsolutePath());
+				}
+			}
+			for (final File lib : libs) {
+				if (!lib.exists()) {
+					this.getLogger().warning(
+							"There was a critical error loading bedwars plugin! Could not find lib: " + lib.getName());
+					Bukkit.getServer().getPluginManager().disablePlugin(this);
+					return;
+				}
+				this.addClassPath(JarUtils.getJarUrl(lib));
+			}
+		} catch (final Exception e) {
+			e.printStackTrace();
+		}
 	}
-	
+
 	private void addClassPath(final URL url) throws IOException {
-        final URLClassLoader sysloader = (URLClassLoader) ClassLoader
-                .getSystemClassLoader();
-        final Class<URLClassLoader> sysclass = URLClassLoader.class;
-        try {
-            final Method method = sysclass.getDeclaredMethod("addURL",
-                    new Class[] { URL.class });
-            method.setAccessible(true);
-            method.invoke(sysloader, new Object[] { url });
-        } catch (final Throwable t) {
-            t.printStackTrace();
-            throw new IOException("Error adding " + url
-                    + " to system classloader");
-        }
-    }
-	
+		final URLClassLoader sysloader = (URLClassLoader) ClassLoader.getSystemClassLoader();
+		final Class<URLClassLoader> sysclass = URLClassLoader.class;
+		try {
+			final Method method = sysclass.getDeclaredMethod("addURL", new Class[] { URL.class });
+			method.setAccessible(true);
+			method.invoke(sysloader, new Object[] { url });
+		} catch (final Throwable t) {
+			t.printStackTrace();
+			throw new IOException("Error adding " + url + " to system classloader");
+		}
+	}
+
 	public DatabaseManager getDatabaseManager() {
 		return this.dbManager;
 	}
@@ -436,7 +437,7 @@ public class Main extends JavaPlugin {
 	public boolean isSpigot() {
 		return this.isSpigot;
 	}
-	
+
 	private boolean getIsSpigot() {
 		try {
 			Package spigotPackage = Package.getPackage("org.spigotmc");
@@ -448,40 +449,40 @@ public class Main extends JavaPlugin {
 		} catch (Exception e) {
 			// nope
 		}
-		
+
 		return false;
 	}
-	
+
 	public int getIntConfig(String key, int defaultInt) {
 		FileConfiguration config = this.getConfig();
-		if(config.contains(key)) {
-			if(config.isInt(key)) {
+		if (config.contains(key)) {
+			if (config.isInt(key)) {
 				return config.getInt(key);
 			}
 		}
-		
+
 		return defaultInt;
 	}
-	
+
 	public String getStringConfig(String key, String defaultString) {
 		FileConfiguration config = this.getConfig();
-		if(config.contains(key)) {
-			if(config.isString(key)) {
+		if (config.contains(key)) {
+			if (config.isString(key)) {
 				return config.getString(key);
 			}
 		}
-		
+
 		return defaultString;
 	}
-	
+
 	public boolean getBooleanConfig(String key, boolean defaultBool) {
 		FileConfiguration config = this.getConfig();
-		if(config.contains(key)) {
-			if(config.isBoolean(key)) {
+		if (config.contains(key)) {
+			if (config.isBoolean(key)) {
 				return config.getBoolean(key);
 			}
 		}
-		
+
 		return defaultBool;
 	}
 
@@ -514,19 +515,13 @@ public class Main extends JavaPlugin {
 		try {
 			if (this.craftbukkit == null) {
 				return Package.getPackage("org.bukkit.craftbukkit."
-						+ Bukkit.getServer().getClass().getPackage().getName()
-								.replace(".", ",").split(",")[3]);
+						+ Bukkit.getServer().getClass().getPackage().getName().replace(".", ",").split(",")[3]);
 			} else {
 				return this.craftbukkit;
 			}
 		} catch (Exception ex) {
-			this.getServer()
-					.getConsoleSender()
-					.sendMessage(
-							ChatWriter.pluginMessage(ChatColor.RED
-									+ Main._l("errors.packagenotfound",
-											ImmutableMap.of("package",
-													"craftbukkit"))));
+			this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(
+					ChatColor.RED + Main._l("errors.packagenotfound", ImmutableMap.of("package", "craftbukkit"))));
 			return null;
 		}
 	}
@@ -535,19 +530,13 @@ public class Main extends JavaPlugin {
 		try {
 			if (this.minecraft == null) {
 				return Package.getPackage("net.minecraft.server."
-						+ Bukkit.getServer().getClass().getPackage().getName()
-								.replace(".", ",").split(",")[3]);
+						+ Bukkit.getServer().getClass().getPackage().getName().replace(".", ",").split(",")[3]);
 			} else {
 				return this.minecraft;
 			}
 		} catch (Exception ex) {
-			this.getServer()
-					.getConsoleSender()
-					.sendMessage(
-							ChatWriter.pluginMessage(ChatColor.RED
-									+ Main._l("errors.packagenotfound",
-											ImmutableMap.of("package",
-													"minecraft server"))));
+			this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(
+					ChatColor.RED + Main._l("errors.packagenotfound", ImmutableMap.of("package", "minecraft server"))));
 			return null;
 		}
 	}
@@ -561,14 +550,8 @@ public class Main extends JavaPlugin {
 
 			return Class.forName(this.craftbukkit.getName() + "." + classname);
 		} catch (Exception ex) {
-			this.getServer()
-					.getConsoleSender()
-					.sendMessage(
-							ChatWriter.pluginMessage(ChatColor.RED
-									+ Main._l("errors.classnotfound",
-											ImmutableMap.of("package",
-													"craftbukkit", "class",
-													classname))));
+			this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.RED
+					+ Main._l("errors.classnotfound", ImmutableMap.of("package", "craftbukkit", "class", classname))));
 			return null;
 		}
 	}
@@ -582,14 +565,8 @@ public class Main extends JavaPlugin {
 
 			return Class.forName(this.minecraft.getName() + "." + classname);
 		} catch (Exception ex) {
-			this.getServer()
-					.getConsoleSender()
-					.sendMessage(
-							ChatWriter.pluginMessage(ChatColor.RED
-									+ Main._l("errors.classnotfound",
-											ImmutableMap.of("package",
-													"minecraft server",
-													"class", classname))));
+			this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.RED + Main
+					._l("errors.classnotfound", ImmutableMap.of("package", "minecraft server", "class", classname))));
 			return null;
 		}
 	}
@@ -604,24 +581,25 @@ public class Main extends JavaPlugin {
 
 		return GameLobbyCountdownRule.getById(id);
 	}
-	
+
 	public boolean metricsEnabled() {
-		if(this.getConfig().contains("plugin-metrics")) {
-			if(this.getConfig().isBoolean("plugin-metrics")) {
+		if (this.getConfig().contains("plugin-metrics")) {
+			if (this.getConfig().isBoolean("plugin-metrics")) {
 				return this.getConfig().getBoolean("plugin-metrics");
 			}
 		}
-		
+
 		return false;
 	}
-	
+
 	public void startMetricsIfEnabled() {
-		if(this.metricsEnabled()) {
+		if (this.metricsEnabled()) {
 			try {
-		        Metrics metrics = new Metrics(this);
-		        metrics.start();
-			} catch(Exception ex) {
-				this.getServer().getConsoleSender().sendMessage(ChatWriter.pluginMessage(ChatColor.RED + "Metrics are enabled, but couldn't send data!"));
+				Metrics metrics = new Metrics(this);
+				metrics.start();
+			} catch (Exception ex) {
+				this.getServer().getConsoleSender().sendMessage(
+						ChatWriter.pluginMessage(ChatColor.RED + "Metrics are enabled, but couldn't send data!"));
 			}
 		}
 	}
@@ -629,25 +607,25 @@ public class Main extends JavaPlugin {
 	public String getFallbackLocale() {
 		return "en";
 	}
-	
+
 	public boolean allPlayersBackToMainLobby() {
-        if(this.getConfig().contains("endgame.all-players-to-mainlobby")) {
-            if(this.getConfig().isBoolean("endgame.all-players-to-mainlobby")) {
-                return this.getConfig().getBoolean("endgame.all-players-to-mainlobby");
-            }
-        }
-        
-        return false;
-    }
-	
+		if (this.getConfig().contains("endgame.all-players-to-mainlobby")) {
+			if (this.getConfig().isBoolean("endgame.all-players-to-mainlobby")) {
+				return this.getConfig().getBoolean("endgame.all-players-to-mainlobby");
+			}
+		}
+
+		return false;
+	}
+
 	public List<String> getAllowedCommands() {
 		FileConfiguration config = this.getConfig();
-		if(config.contains("allowed-commands")) {
-			if(config.isList("allowed-commands")) {
+		if (config.contains("allowed-commands")) {
+			if (config.isList("allowed-commands")) {
 				return config.getStringList("allowed-commands");
 			}
 		}
-		
+
 		return new ArrayList<String>();
 	}
 
@@ -664,20 +642,19 @@ public class Main extends JavaPlugin {
 		new BlockListener();
 		new PlayerListener();
 		EntityListener el = new EntityListener();
-		if(Utils.isSupportingTitles()) {
-		    new Entity18Listener(el);
+		if (Utils.isSupportingTitles()) {
+			new Entity18Listener(el);
 		}
-		
+		new HangingListener();
 		new ServerListener();
 		new SignListener();
 		new ChunkListener();
-		
+
 		SpecialItem.loadSpecials();
 	}
 
 	private void registerConfigurationClasses() {
-		ConfigurationSerialization.registerClass(RessourceSpawner.class,
-				"RessourceSpawner");
+		ConfigurationSerialization.registerClass(RessourceSpawner.class, "RessourceSpawner");
 		ConfigurationSerialization.registerClass(Team.class, "Team");
 	}
 
@@ -721,8 +698,7 @@ public class Main extends JavaPlugin {
 		return this.commands;
 	}
 
-	private ArrayList<BaseCommand> filterCommandsByPermission(
-			ArrayList<BaseCommand> commands, String permission) {
+	private ArrayList<BaseCommand> filterCommandsByPermission(ArrayList<BaseCommand> commands, String permission) {
 		Iterator<BaseCommand> it = commands.iterator();
 
 		while (it.hasNext()) {
@@ -737,8 +713,7 @@ public class Main extends JavaPlugin {
 
 	@SuppressWarnings("unchecked")
 	public ArrayList<BaseCommand> getBaseCommands() {
-		ArrayList<BaseCommand> commands = (ArrayList<BaseCommand>) this.commands
-				.clone();
+		ArrayList<BaseCommand> commands = (ArrayList<BaseCommand>) this.commands.clone();
 		commands = this.filterCommandsByPermission(commands, "base");
 
 		return commands;
@@ -746,13 +721,12 @@ public class Main extends JavaPlugin {
 
 	@SuppressWarnings("unchecked")
 	public ArrayList<BaseCommand> getSetupCommands() {
-		ArrayList<BaseCommand> commands = (ArrayList<BaseCommand>) this.commands
-				.clone();
+		ArrayList<BaseCommand> commands = (ArrayList<BaseCommand>) this.commands.clone();
 		commands = this.filterCommandsByPermission(commands, "setup");
 
 		return commands;
 	}
-	
+
 	@SuppressWarnings("unchecked")
 	public ArrayList<BaseCommand> getCommandsByPermission(String permission) {
 		ArrayList<BaseCommand> commands = (ArrayList<BaseCommand>) this.commands.clone();
@@ -766,33 +740,28 @@ public class Main extends JavaPlugin {
 	}
 
 	private void startTimeListener() {
-		this.timeTask = this.getServer().getScheduler()
-				.runTaskTimer(this, new Runnable() {
+		this.timeTask = this.getServer().getScheduler().runTaskTimer(this, new Runnable() {
 
-					@Override
-					public void run() {
-						for (Game g : Main.getInstance().getGameManager()
-								.getGames()) {
-							if (g.getState() == GameState.RUNNING) {
-								g.getRegion().getWorld().setTime(g.getTime());
-							}
-						}
+			@Override
+			public void run() {
+				for (Game g : Main.getInstance().getGameManager().getGames()) {
+					if (g.getState() == GameState.RUNNING) {
+						g.getRegion().getWorld().setTime(g.getTime());
 					}
-				}, (long) 5 * 20, (long) 5 * 20);
+				}
+			}
+		}, (long) 5 * 20, (long) 5 * 20);
 	}
-	
+
 	public static String _l(String localeKey, String singularValue, Map<String, String> params) {
-		if(params.get(singularValue).equals("1")){
-			return (String) Main.getInstance().getLocalization()
-					.get(localeKey + "-one", params);
+		if (params.get(singularValue).equals("1")) {
+			return (String) Main.getInstance().getLocalization().get(localeKey + "-one", params);
 		}
-		return (String) Main.getInstance().getLocalization()
-				.get(localeKey, params);
+		return (String) Main.getInstance().getLocalization().get(localeKey, params);
 	}
 
 	public static String _l(String localeKey, Map<String, String> params) {
-		return (String) Main.getInstance().getLocalization()
-				.get(localeKey, params);
+		return (String) Main.getInstance().getLocalization().get(localeKey, params);
 	}
 
 	public static String _l(String localeKey) {
@@ -805,18 +774,17 @@ public class Main extends JavaPlugin {
 		} catch (Exception ex) {
 			// Timer isn't running. Just ignore.
 		}
-		
+
 		try {
 			this.updateChecker.cancel();
-		} catch(Exception ex) {
-		 // Timer isn't running. Just ignore.
+		} catch (Exception ex) {
+			// Timer isn't running. Just ignore.
 		}
 	}
 
 	public void reloadLocalization() {
 		this.localization.saveLocales(false);
-		this.localization.loadLocale(this.getConfig().getString("locale"),
-				false);
+		this.localization.loadLocale(this.getConfig().getString("locale"), false);
 	}
 
 	public boolean spectationEnabled() {
@@ -855,38 +823,38 @@ public class Main extends JavaPlugin {
 
 	public Integer getRespawnProtectionTime() {
 		FileConfiguration config = this.getConfig();
-		if(config.contains("respawn-protection")) {
-			if(config.isInt("respawn-protection")) {
+		if (config.contains("respawn-protection")) {
+			if (config.isInt("respawn-protection")) {
 				return config.getInt("respawn-protection");
 			}
 		}
-		
+
 		return 0;
 	}
-	
+
 	public boolean isLocationSerializable() {
-		if(Main.locationSerializable == null) {
+		if (Main.locationSerializable == null) {
 			try {
 				Location.class.getMethod("serialize");
 				Main.locationSerializable = true;
-			} catch(Exception ex) {
+			} catch (Exception ex) {
 				Main.locationSerializable = false;
 			}
 		}
-		
+
 		return Main.locationSerializable;
 	}
 
 	public FileConfiguration getShopConfig() {
 		return this.shopConfig;
 	}
-	
+
 	public boolean isHologramsEnabled() {
-        return this.getServer().getPluginManager().isPluginEnabled("HolographicDisplays");
-    }
-	
-    public HolographicDisplaysInteraction getHolographicInteractor() {
-        return this.holographicInteraction;
-    }
+		return this.getServer().getPluginManager().isPluginEnabled("HolographicDisplays");
+	}
+
+	public HolographicDisplaysInteraction getHolographicInteractor() {
+		return this.holographicInteraction;
+	}
 
 }

